File: scripts/investor_returns.py
"""Stacks investor YTD returns — daily precompute for the 13F investor hub.

portfolios.json only refreshes quarterly (13f-refresh.yml, cron
"0 22 8-19 2,5,8,11 *" — Feb/May/Aug/Nov only) and carries no market-price
data at all, so it cannot back a daily-changing "YTD return" column on its
own. This script fills that gap: once a day it prices each investor's
*current* 13F holdings against live quotes and writes a small summary,
data/investor-returns.json, for the hub UI to read directly. It does not
touch index.html or any other UI file.

Eligibility rule — copied verbatim from the compare screen's
invComputeValueSeries() (index.html, ~line 7130-7141), NOT reinvented:

    sourceHoldings = inv.all_holdings || inv.holdings || []
    reals    = sourceHoldings.filter(h => h.change !== "exit"
                                        && h.put_call !== "PUT"
                                        && h.put_call !== "CALL")
    eligible = reals.filter(h => h.ticker)

See eligible_holdings() below for the direct Python port.

Price source — the repo's own Cloudflare Worker, same one the frontend
calls (COMMENTS_API in index.html, resolved here via scripts/worker_url.py,
which normalizes to the canonical https://api.stacksdaily.com):

    GET {worker}/quote?s=<ticker>&r=1y  ->  {t: [epoch...], closes: [...], dates: [...]}

One ticker per request (the worker's `s` sanitizer strips commas, so
batching is not possible — see worker/index.js). Tickers are deduplicated
across all 16 investors first, so each ticker is fetched at most once per
run, then fetched with a small thread pool (see CONCURRENCY below).

YTD math, per investor, over only the *priced* eligible holdings:

    value(t)  = sum(shares * close_at_or_before(t))   for each priced holding
    ytd_pct   = value(latest) / value(first trading day on/after Jan 1) - 1

"latest"/"first trading day" both come off ONE canonical daily calendar (the
longest series among everything fetched this run) rather than a
per-investor calendar. Because every 13F holding is, by definition, a
US-listed reportable security (fetch_13f.py's normalize_ticker() always
appends ".us"), every ticker here trades on the same NYSE/NASDAQ calendar,
so a single shared calendar and a per-investor "pick the longest series I
hold" calendar (what the frontend actually does) agree on every trading
date in practice. This keeps `ytd_start` a single, reportable top-level
field instead of 16 slightly different ones.

Coverage gate — coverage_pct is the fraction of an investor's ELIGIBLE 13F
`value` (the dollar amount from the filing, summed over the same eligible
list above) whose ticker produced a usable price series this run. If
coverage_pct is below COVERAGE_THRESHOLD, ytd_pct is set to null (not 0, not
omitted) so the UI can render "insufficient data" instead of a misleading
number computed from a fraction of the real portfolio. See COVERAGE_THRESHOLD
below for the number and why.

Failure handling, matching fetch_13f.py's philosophy (one bad ticker must
not take down the run) plus one addition this file specifically needs
(silently blanking a previously-good column is worse than doing nothing):

  - One ticker's /quote request failing (timeout, HTTP error, no data) only
    drops that ticker; every other ticker and every other investor are
    unaffected. Failed tickers are logged with a reason.
  - If a run raises an uncaught exception, nothing is written — the
    existing data/investor-returns.json (whatever the workflow checked out)
    stays exactly as it was, so `git diff` sees no change and the calling
    workflow's "commit only if changed" step naturally does not commit.
  - A SMALL drop in usable investors (coverage_pct dipping under
    COVERAGE_THRESHOLD for one or two names - e.g. one ticker the Worker/
    Yahoo temporarily has no chart for) is normal day-to-day noise, not a
    pipeline failure: it blanks one or two cells to null, which the UI
    already renders honestly as "no data" alongside that investor's
    coverage_pct. The file IS written in this case (with those investors'
    ytd_pct: null), a warning is logged, and the run exits 0. Refusing to
    write over a single flaky ticker would wedge the whole daily pipeline
    and (via the calling workflow's failure() step) open a fresh GitHub
    issue every single day forever - worse than the one blank cell it was
    trying to prevent.
  - Only a SYSTEMIC drop - usable count falling by more than
    REGRESSION_TOLERANCE investors versus the previously written file - is
    treated as the "do not touch the file" case: log the regression
    loudly, do NOT write, and exit non-zero, so a stale "yesterday" file
    (via the calling workflow's "commit only if changed" step never
    running) beats today's, largely-blanked one.
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import time
import urllib.error
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime, timezone

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.dirname(HERE)
sys.path.insert(0, HERE)
import worker_url  # noqa: E402 - needs sys.path tweak above

PORTFOLIOS_PATH = os.path.join(ROOT, "portfolios.json")
OUTPUT_PATH = os.path.join(ROOT, "data", "investor-returns.json")

WORKER = worker_url.worker_base()
UA = "Stacks/1.0 (stacksdaily.com; investor-returns; contact@stacksdaily.com)"

# Modest concurrency for a batch job hitting our own edge-cached Worker (not
# Yahoo directly). The frontend itself runs 5-way concurrent (invMapLimit(...,
# 5, ...)) against the same endpoint for a single investor's holdings; this
# run fans out across the whole 16-investor roster's unique tickers in one
# go, so a slightly higher pool (8) trades a bit more parallel load for a
# materially shorter wall-clock run, while staying well short of anything
# that would look like abuse to Yahoo behind the Worker.
CONCURRENCY = 8
REQUEST_TIMEOUT = 15  # seconds, per HTTP attempt
MAX_RETRIES = 3
RETRY_BASE_SLEEP = 1.5  # seconds; attempt N sleeps RETRY_BASE_SLEEP * N

# Below this fraction of an investor's eligible 13F `value` priced, the YTD
# number would be computed from too small a slice of the real portfolio to
# be worth showing as a number - report coverage (so the UI can say "partial
# data") but null out ytd_pct rather than publish something misleading.
# Chosen at 0.90: a single very large, unpriceable position (e.g. an OTC or
# newly-listed ticker the Worker/Yahoo has no chart for) can easily account
# for 5-10% of one investor's book without being unusual, so a much stricter
# gate (e.g. 0.99) would flip that investor to "no data" on ordinary market
# noise. Below 90% priced, more than a tenth of the book is invisible to the
# math, which is large enough that a misleading ytd_pct is a real risk.
COVERAGE_THRESHOLD = 0.90

# How many FEWER usable investors than the previously written file counts as
# a "systemic" regression worth refusing to write over (see module
# docstring). One or two investors dipping under COVERAGE_THRESHOLD on a
# given day is ordinary noise (a single ticker the Worker/Yahoo temporarily
# has no chart for) and is written anyway with a warning - the UI already
# renders a null ytd_pct as "no data" next to that investor's coverage_pct,
# so one blank cell is not worth wedging the whole daily pipeline and
# opening a fresh issue over. A drop bigger than this (out of 16 investors
# total) stops looking like one flaky ticker and starts looking like the
# Worker itself being down or broken, which IS worth refusing to publish.
REGRESSION_TOLERANCE = 2


def eligible_holdings(inv: dict) -> list:
    """Direct port of invComputeValueSeries' eligibility filter
    (index.html, ~line 7138-7141). Do not change this rule here without
    changing it there first - the two must always agree.
    """
    source = inv.get("all_holdings") or inv.get("holdings") or []
    reals = [
        h for h in source
        if h and h.get("change") != "exit"
        and h.get("put_call") not in ("PUT", "CALL")
    ]
    return [h for h in reals if h.get("ticker")]


def load_json(path: str, default):
    try:
        with open(path, encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return default


def http_get_json(url: str, retries: int = MAX_RETRIES) -> dict:
    req = urllib.request.Request(url, headers={"User-Agent": UA, "Accept-Encoding": "identity"})
    last_err = None
    for attempt in range(retries):
        try:
            with urllib.request.urlopen(req, timeout=REQUEST_TIMEOUT) as r:
                return json.loads(r.read())
        except urllib.error.HTTPError as e:
            last_err = e
            if e.code in (400, 404):
                # Worker itself rejected the symbol or had no data - retrying
                # will not help.
                raise
            time.sleep(RETRY_BASE_SLEEP * (attempt + 1))
        except Exception as e:  # noqa: BLE001 - network/timeout, worth a retry
            last_err = e
            time.sleep(RETRY_BASE_SLEEP * (attempt + 1))
    raise last_err


def fetch_ticker_series(ticker: str):
    """Return (ticker, series_dict_or_None, error_str_or_None)."""
    url = "%s/quote?s=%s&r=1y" % (WORKER, ticker)
    try:
        j = http_get_json(url)
    except Exception as e:  # noqa: BLE001 - reported to caller, not raised
        return ticker, None, "%s: %s" % (type(e).__name__, e)
    if not j or j.get("error"):
        return ticker, None, "worker returned error: %s" % j.get("error") if j else "empty response"
    ts, closes = j.get("t"), j.get("closes")
    if not ts or not closes or len(closes) < 2:
        return ticker, None, "no usable series (< 2 closes)"
    return ticker, j, None


def value_at(series: dict, t: int):
    """Port of invValueAt(): last close at-or-before t, clamped to the
    series' own edges. `series` here uses the worker's own field names
    (t/closes), same as the frontend's `q`.
    """
    ts, cs = series.get("t"), series.get("closes")
    if not ts:
        return None
    if t <= ts[0]:
        return cs[0]
    if t >= ts[-1]:
        return cs[-1]
    lo, hi = 0, len(ts) - 1
    while lo < hi:
        mid = (lo + hi + 1) // 2
        if ts[mid] <= t:
            lo = mid
        else:
            hi = mid - 1
    return cs[lo]


def fetch_all_series(tickers: list, concurrency: int = CONCURRENCY):
    series_by_ticker = {}
    errors = {}
    with ThreadPoolExecutor(max_workers=concurrency) as ex:
        futures = {ex.submit(fetch_ticker_series, t): t for t in tickers}
        for fut in as_completed(futures):
            ticker, series, err = fut.result()
            if series is not None:
                series_by_ticker[ticker] = series
            else:
                errors[ticker] = err
    return series_by_ticker, errors


def compute_investor(inv: dict, series_by_ticker: dict, start_t: int, end_t: int):
    elig = eligible_holdings(inv)
    eligible_count = len(elig)
    eligible_value = sum(h.get("value") or 0 for h in elig)
    priced = [h for h in elig if h.get("ticker") in series_by_ticker]
    priced_count = len(priced)
    priced_value = sum(h.get("value") or 0 for h in priced)
    coverage_pct = round(priced_value / eligible_value, 4) if eligible_value else 0.0

    ytd_pct = None
    if priced and coverage_pct >= COVERAGE_THRESHOLD:
        value_start = 0.0
        value_end = 0.0
        for h in priced:
            s = series_by_ticker[h["ticker"]]
            shares = h.get("shares") or 0
            p0 = value_at(s, start_t)
            p1 = value_at(s, end_t)
            if p0 is not None:
                value_start += p0 * shares
            if p1 is not None:
                value_end += p1 * shares
        if value_start > 0:
            ytd_pct = round(value_end / value_start - 1, 4)

    return {
        "ytd_pct": ytd_pct,
        "coverage_pct": coverage_pct,
        "priced_holdings": priced_count,
        "eligible_holdings": eligible_count,
    }, elig


def usable_count(investors_map: dict) -> int:
    return sum(1 for v in investors_map.values() if isinstance(v, dict) and v.get("ytd_pct") is not None)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--dry-run", action="store_true",
        help="compute and print, but never write data/investor-returns.json",
    )
    args = parser.parse_args()

    portfolios = load_json(PORTFOLIOS_PATH, {"investors": []})
    investors = portfolios.get("investors", [])
    if not investors:
        print("[fail] portfolios.json has no investors - nothing to do", file=sys.stderr)
        return 1

    per_investor_eligible = {}
    all_tickers = set()
    for inv in investors:
        elig = eligible_holdings(inv)
        per_investor_eligible[inv["slug"]] = elig
        for h in elig:
            all_tickers.add(h["ticker"])
    tickers = sorted(all_tickers)
    print("[info] investors=%d unique_tickers=%d" % (len(investors), len(tickers)))

    t0 = time.time()
    series_by_ticker, fetch_errors = fetch_all_series(tickers)
    elapsed = time.time() - t0
    print("[info] fetched %d/%d tickers ok in %.1fs (concurrency=%d, %d failed)"
          % (len(series_by_ticker), len(tickers), elapsed, CONCURRENCY, len(fetch_errors)))
    for t in sorted(fetch_errors):
        print("[warn] ticker failed: %s (%s)" % (t, fetch_errors[t]))

    if not series_by_ticker:
        print("[fail] every ticker fetch failed - refusing to write a zero-coverage file", file=sys.stderr)
        return 1

    # Canonical daily calendar: the longest series fetched this run. See
    # module docstring for why one shared calendar is fine here (every
    # ticker is a .us-suffixed 13F holding, i.e. the same NYSE/NASDAQ
    # trading calendar).
    best_ticker = max(series_by_ticker, key=lambda t: len(series_by_ticker[t]["t"]))
    best = series_by_ticker[best_ticker]
    calendar, dates = best["t"], best.get("dates") or []
    now = datetime.now(timezone.utc)
    year_start_epoch = int(datetime(now.year, 1, 1, tzinfo=timezone.utc).timestamp())
    start_idx = None
    for i, t in enumerate(calendar):
        if t >= year_start_epoch:
            start_idx = i
            break
    if start_idx is None:
        start_idx = len(calendar) - 1
    start_t, end_t = calendar[start_idx], calendar[-1]
    ytd_start = dates[start_idx] if start_idx < len(dates) else \
        datetime.fromtimestamp(start_t, tz=timezone.utc).date().isoformat()
    print("[info] canonical calendar from %s: ytd_start=%s (%d points)"
          % (best_ticker, ytd_start, len(calendar)))

    new_investors = {}
    for inv in investors:
        slug = inv["slug"]
        result, elig = compute_investor(inv, series_by_ticker, start_t, end_t)
        new_investors[slug] = result
        cov_str = "%.1f%%" % (result["coverage_pct"] * 100)
        if result["ytd_pct"] is not None:
            print("[ok] %s: ytd=%.2f%% coverage=%s priced=%d/%d"
                  % (slug, result["ytd_pct"] * 100, cov_str,
                     result["priced_holdings"], result["eligible_holdings"]))
        else:
            print("[warn] %s: ytd=null coverage=%s (below %.0f%% threshold or no priced holdings) priced=%d/%d"
                  % (slug, cov_str, COVERAGE_THRESHOLD * 100,
                     result["priced_holdings"], result["eligible_holdings"]))

    new_usable = usable_count(new_investors)
    prev = load_json(OUTPUT_PATH, None)
    prev_usable = usable_count(prev.get("investors", {})) if prev else 0
    print("[info] usable ytd_pct count: new=%d previous=%d" % (new_usable, prev_usable))

    if prev is not None and new_usable < prev_usable - REGRESSION_TOLERANCE:
        print("::error::investor-returns systemic regression: usable count dropped from %d to %d "
              "(more than the %d-investor tolerance) - refusing to overwrite %s (previous file kept)"
              % (prev_usable, new_usable, REGRESSION_TOLERANCE, OUTPUT_PATH), file=sys.stderr)
        return 1
    if prev is not None and new_usable < prev_usable:
        print("::warning::investor-returns: usable count dipped from %d to %d "
              "(within the %d-investor tolerance) - writing anyway, affected investors show ytd_pct: null"
              % (prev_usable, new_usable, REGRESSION_TOLERANCE))

    out = {
        "as_of": now.isoformat(),
        "ytd_start": ytd_start,
        "investors": new_investors,
    }

    if args.dry_run:
        print("[info] --dry-run: not writing %s" % OUTPUT_PATH)
        print(json.dumps(out, indent=1, ensure_ascii=False))
        return 0

    os.makedirs(os.path.dirname(OUTPUT_PATH), exist_ok=True)
    with open(OUTPUT_PATH, "w", encoding="utf-8") as f:
        json.dump(out, f, indent=1, ensure_ascii=False)
        f.write("\n")
    print("[info] wrote %s" % OUTPUT_PATH)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/investor_returns.mjs
// Stacks investor YTD returns — daily precompute for the 13F investor hub,
// v2: headless-browser producer, NOT a Python reimplementation.
//
// WHY THIS EXISTS INSTEAD OF A PYTHON PORT (read before "optimizing" this
// back into one):
//
// The previous version of this file was scripts/investor_returns.py, a
// line-by-line JavaScript->Python port of index.html's
// invComputeHistoricalValueSeries / invComputeValueSeries / the YTD window
// slice in invPaintValueChart. It was carefully annotated, reviewed, and
// still drifted from the live compare screen's own "연중 수익률" number for
// 2 of the 16 investors in the roster — situational-awareness by 2.7pp and
// tiger-global by 1.0pp — for a reason that could not be pinned down by
// further code comparison (float edge cases in the quarter-chaining
// rescale loop, subtle fetch/ordering differences, or something else
// entirely; it was never isolated). Every other investor matched exactly.
//
// A hub column that is off by single-digit points of "pp" for a couple of
// names, in a way nobody can explain, is worse than useless — it is a
// second, silently-different definition of the same on-screen label. So
// instead of chasing the remaining drift in a reimplementation, this
// script now drives a real headless Chromium against the live site and
// calls the site's OWN window.invComputeValueSeries()/window.quote1y()
// functions directly. There is exactly one implementation of the YTD math
// in this project (index.html); this script is a caller of it, not a
// second copy of it. If invComputeValueSeries ever changes, this script
// picks up the change automatically — nothing here needs to be re-ported.
//
// portfolios.json only refreshes quarterly (13f-refresh.yml) and carries
// no market-price data at all, so it cannot back a daily-changing "YTD
// return" column on its own — that gap is what this script fills, once a
// day, by asking the live page to do the same computation it does for a
// human visitor and writing the result to data/investor-returns.json for
// the hub UI to read directly. It does not touch index.html or
// portfolios.json.
//
// Failure handling (ported in spirit, not in code, from the previous
// Python version — see its git history for the fuller reasoning this
// summarizes):
//   - One investor's computation failing (network hiccup, unexpected
//     shape, no priceable tickers) only blanks that investor; every other
//     investor is unaffected.
//   - A SMALL drop in "usable" investors (coverage dipping under
//     COVERAGE_THRESHOLD for one or two names on an ordinary day) is
//     written anyway, with those investors' ytd_pct: null — the UI already
//     renders that honestly next to the investor's coverage_pct.
//   - A drop bigger than REGRESSION_TOLERANCE, or the site/function not
//     being reachable/found at all, or fewer than MIN_INVESTORS_WITH_RESULT
//     investors resolving anything, is treated as systemic: refuse to
//     write, exit non-zero, and let yesterday's file stand.
//   - Every successful investor's YTD window must start on the SAME
//     calendar date (they all key off "Jan 1 of the year of the latest
//     shared trading day" the same way) — if they disagree, something is
//     structurally wrong with the run, so abort rather than publish a
//     column where each cell silently measures a different window.
//
// Note on eligible/priced holdings counts: window.invComputeValueSeries()
// does not itself return holding counts (only a weighted `coverage`
// fraction), so this script derives eligible/priced independently, but
// EXACTLY the way the frontend does it, not by approximation from the
// coverage weight:
//   eligible = rows in (all_holdings || holdings) of the investor's
//              CURRENT (latest) disclosed book where
//              change !== "exit" && put_call !== "PUT" && put_call !== "CALL"
//              (mirrors `currentRows` in invComputeHistoricalValueSeries,
//              index.html; no ticker requirement — an unpriceable
//              position still counts against eligibility/coverage).
//   priced   = of those, rows with a ticker AND a numeric `shares` count
//              AND for which window.quote1y(ticker) resolves to a series
//              with more than one point and no `.error` (the exact
//              condition invComputeHistoricalValueSeries/invComputeValueSeries
//              use to decide whether a ticker's quote is usable:
//              `r.q && r.q.t && r.q.t.length > 1`). Calling window.quote1y
//              directly for these tickers reuses the SAME in-page cache
//              invComputeValueSeries already populated, so this costs no
//              extra network round trips for tickers already fetched.
//
// This script is a build tool, not the ES5-constrained runtime code in
// assets/investor-compare.js — modern Node ESM (top-level await, optional
// chaining, etc.) is fine here.

import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { chromium } from "@playwright/test";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.dirname(HERE);
const OUTPUT_PATH = path.join(ROOT, "data", "investor-returns.json");

const SITE_URL = "https://stacksdaily.com/";
const READY_TIMEOUT_MS = 60_000; // window.invComputeValueSeries/window.quote1y must appear within this
const NAV_TIMEOUT_MS = 45_000;

// Same constants/intent as the retired scripts/investor_returns.py — see
// that file's git history for the full reasoning; summarized above.
export const COVERAGE_THRESHOLD = 0.90;
export const REGRESSION_TOLERANCE = 2;
export const MIN_INVESTORS_WITH_RESULT = 10;

// --------------------------------------------------------------------------
// Pure helpers — no browser, no network. Unit-testable via dynamic import.
// --------------------------------------------------------------------------

/** Mirrors invPaintValueChart's ceiling search (index.html):
 *    var startIdx = 0;
 *    while (startIdx < calendar.length - 1 && calendar[startIdx] < requestedStart) startIdx++;
 * calendar entries are epoch seconds (numbers). */
export function ytdStartIndex(calendar, requestedStartEpoch) {
  let i = 0;
  while (i < calendar.length - 1 && calendar[i] < requestedStartEpoch) i++;
  return i;
}

/** UTC calendar date (YYYY-MM-DD) of an epoch-seconds timestamp. */
export function isoDateUTC(epochSeconds) {
  return new Date(epochSeconds * 1000).toISOString().slice(0, 10);
}

/** Same rounding the retired Python version used: 4 decimal places on the
 * fraction (e.g. 0.0902 == +9.02%). */
export function round4(x) {
  return x === null || x === undefined ? null : Math.round(x * 10000) / 10000;
}

/** Count investors with a non-null ytd_pct in the {slug: {...}} map. */
export function usableCount(investorsMap) {
  return Object.values(investorsMap || {}).filter(
    (v) => v && v.ytd_pct !== null && v.ytd_pct !== undefined
  ).length;
}

/** COVERAGE_THRESHOLD gate: below it, null out ytd_pct but keep coverage_pct
 * so the UI can explain why. Applied identically regardless of how the raw
 * result was produced. */
export function applyCoverageGate(result) {
  if (result.ytd_pct !== null && result.coverage_pct < COVERAGE_THRESHOLD) {
    return { ...result, ytd_pct: null };
  }
  return result;
}

/** Regression guard: refuse to publish if usable count fell by more than
 * REGRESSION_TOLERANCE vs. the previously written file. `prevUsable` is
 * null when there is no previous file (first run — never a regression). */
export function checkRegression(newUsable, prevUsable) {
  if (prevUsable === null || prevUsable === undefined) return { ok: true, warn: false };
  if (newUsable < prevUsable - REGRESSION_TOLERANCE) {
    return {
      ok: false,
      message:
        `systemic regression: usable count dropped from ${prevUsable} to ${newUsable} ` +
        `(more than the ${REGRESSION_TOLERANCE}-investor tolerance) - refusing to overwrite ` +
        `${OUTPUT_PATH} (previous file kept)`,
    };
  }
  return {
    ok: true,
    warn: newUsable < prevUsable,
    message:
      newUsable < prevUsable
        ? `usable count dipped from ${prevUsable} to ${newUsable} (within the ` +
          `${REGRESSION_TOLERANCE}-investor tolerance) - writing anyway, affected investors show ytd_pct: null`
        : null,
  };
}

/** Format "now" as an ISO-8601 UTC timestamp with a "+00:00" offset (rather
 * than "Z") and 6 fractional digits, matching the shape the retired Python
 * version's datetime.now(timezone.utc).isoformat() produced (e.g.
 * "2026-08-12T07:28:29.296472+00:00"). Node's Date only carries millisecond
 * precision, so the last 3 digits are zero-padded rather than genuine
 * microseconds — the FORMAT matches exactly, the precision does not (see
 * report for this caveat). */
export function nowIsoUtcMicros(date = new Date()) {
  const iso = date.toISOString(); // e.g. 2026-08-12T07:28:29.296Z
  const [head, msZ] = iso.split(".");
  const ms = msZ.slice(0, 3);
  return `${head}.${ms}000+00:00`;
}

async function loadJson(filePath, fallback) {
  try {
    return JSON.parse(await fs.readFile(filePath, "utf8"));
  } catch {
    return fallback;
  }
}

// --------------------------------------------------------------------------
// In-page computation. Runs inside the live site via page.evaluate() so it
// calls the site's OWN window.invComputeValueSeries()/window.quote1y() —
// see the file-level comment for why this is deliberate.
// --------------------------------------------------------------------------

/**
 * Runs in the browser page context for a single investor. `inv` is the raw
 * investor object out of /portfolios.json (must be JSON-serializable to
 * cross the page.evaluate boundary). Returns a plain, JSON-serializable
 * result object; never throws across the boundary (network/shape errors
 * are caught and reported as { ok:false, reason }).
 */
/* eslint-disable no-undef -- this function body executes inside the page */
async function computeInvestorInPage(inv) {
  try {
    const res = await window.invComputeValueSeries(inv);
    if (!res || !res.calendar || !res.values || res.calendar.length < 2) {
      return { ok: false, reason: "no usable calendar/values from invComputeValueSeries" };
    }
    const calendar = res.calendar;
    const values = res.values;

    // Mirror invPaintValueChart's ytd branch exactly (index.html):
    //   var end = calendar[calendar.length - 1], endDate = new Date(end * 1000);
    //   requestedStart = Date.UTC(endDate.getUTCFullYear(), 0, 1) / 1000;
    //   var startIdx = 0;
    //   while (startIdx < calendar.length - 1 && calendar[startIdx] < requestedStart) startIdx++;
    const endEpoch = calendar[calendar.length - 1];
    const endYear = new Date(endEpoch * 1000).getUTCFullYear();
    const requestedStart = Date.UTC(endYear, 0, 1) / 1000;
    let startIdx = 0;
    while (startIdx < calendar.length - 1 && calendar[startIdx] < requestedStart) startIdx++;

    const startVal = values[startIdx];
    const endVal = values[values.length - 1];
    if (!(typeof startVal === "number" && startVal > 0)) {
      return { ok: false, reason: "ytd start value is not usable (<=0 or missing)" };
    }
    const ytd_pct = endVal / startVal - 1;
    const ytd_start_epoch = calendar[startIdx];
    const coverage_pct = typeof res.coverage === "number" ? res.coverage : 0;

    // eligible/priced from the CURRENT book — see file header for the exact
    // rule this mirrors (currentRows in invComputeHistoricalValueSeries /
    // rawTotal's skip condition).
    const source = (inv && (inv.all_holdings || inv.holdings)) || [];
    const reals = source.filter(
      (h) => h && h.change !== "exit" && h.put_call !== "PUT" && h.put_call !== "CALL"
    );
    const priceableCandidates = reals.filter(
      (h) => h && h.ticker && typeof h.shares === "number"
    );
    const tickers = Array.from(new Set(priceableCandidates.map((h) => h.ticker)));
    const checks = await Promise.all(
      tickers.map(async (t) => {
        try {
          const q = await window.quote1y(t);
          const usable = !!(q && !q.error && q.t && q.t.length > 1);
          return [t, usable];
        } catch {
          return [t, false];
        }
      })
    );
    const pricedTickers = new Set(checks.filter(([, ok]) => ok).map(([t]) => t));
    const priced_holdings = priceableCandidates.filter((h) => pricedTickers.has(h.ticker)).length;
    const eligible_holdings = reals.length;

    return {
      ok: true,
      ytd_pct,
      coverage_pct,
      ytd_start_epoch,
      priced_holdings,
      eligible_holdings,
    };
  } catch (e) {
    return { ok: false, reason: String((e && e.message) || e) };
  }
}
/* eslint-enable no-undef */

// --------------------------------------------------------------------------
// main
// --------------------------------------------------------------------------

async function main() {
  const prev = await loadJson(OUTPUT_PATH, null);
  const prevUsable = prev ? usableCount(prev.investors) : null;

  const browser = await chromium.launch({ headless: true });
  // NOTE: process.exitCode is set directly at each failure point below,
  // not via a local variable read after the try/finally - a `return`
  // inside `try` runs `finally` and then exits the function immediately,
  // so any statement placed after the try/finally block would never run
  // on those paths. Setting process.exitCode inline avoids that trap.
  try {
    const page = await browser.newPage();
    page.setDefaultTimeout(NAV_TIMEOUT_MS);
    console.log(`[info] navigating to ${SITE_URL}`);
    await page.goto(SITE_URL, { waitUntil: "domcontentloaded", timeout: NAV_TIMEOUT_MS });

    try {
      await page.waitForFunction(
        () =>
          typeof window.invComputeValueSeries === "function" &&
          typeof window.quote1y === "function",
        { timeout: READY_TIMEOUT_MS }
      );
    } catch {
      console.error(
        `[fail] window.invComputeValueSeries/window.quote1y did not appear within ${READY_TIMEOUT_MS}ms - ` +
          `site may be down or restructured; refusing to write ${OUTPUT_PATH}`
      );
      process.exitCode = 1;
      return;
    }

    const portfolios = await page.evaluate(async () => {
      const r = await fetch("/portfolios.json");
      if (!r.ok) throw new Error(`portfolios.json fetch failed: HTTP ${r.status}`);
      return r.json();
    });
    const investors = (portfolios && portfolios.investors) || [];
    if (!investors.length) {
      console.error("[fail] /portfolios.json has no investors - nothing to do");
      process.exitCode = 1;
      return;
    }
    console.log(`[info] investors=${investors.length}`);

    const perInvestor = {}; // slug -> in-page result (ok:true|false)
    for (const inv of investors) {
      const slug = inv.slug;
      let result;
      try {
        result = await page.evaluate(computeInvestorInPage, inv);
      } catch (e) {
        result = { ok: false, reason: `page.evaluate threw: ${(e && e.message) || e}` };
      }
      perInvestor[slug] = result;
      if (result.ok) {
        console.log(
          `[ok] ${slug}: ytd=${(result.ytd_pct * 100).toFixed(2)}% coverage=${(result.coverage_pct * 100).toFixed(1)}% ` +
            `priced=${result.priced_holdings}/${result.eligible_holdings}`
        );
      } else {
        console.log(`[warn] ${slug}: no result (${result.reason})`);
      }
    }

    const successful = Object.entries(perInvestor).filter(([, r]) => r.ok);
    if (successful.length < MIN_INVESTORS_WITH_RESULT) {
      console.error(
        `[fail] only ${successful.length}/${investors.length} investors produced any result ` +
          `(minimum ${MIN_INVESTORS_WITH_RESULT}) - refusing to write ${OUTPUT_PATH}`
      );
      process.exitCode = 1;
      return;
    }

    // Every successful investor must resolve to the SAME ytd_start date -
    // see file header. Disagreement means something structural is wrong
    // (e.g. different investors landing on different "latest calendar
    // point" years), not ordinary per-investor noise.
    const startDatesBySlug = {};
    for (const [slug, r] of successful) {
      startDatesBySlug[slug] = isoDateUTC(r.ytd_start_epoch);
    }
    const distinctDates = Array.from(new Set(Object.values(startDatesBySlug)));
    if (distinctDates.length > 1) {
      console.error(
        `[fail] investors disagree on ytd_start date - refusing to write ${OUTPUT_PATH}:\n` +
          JSON.stringify(startDatesBySlug, null, 1)
      );
      process.exitCode = 1;
      return;
    }
    const ytdStart = distinctDates[0];

    const newInvestors = {};
    const gated = {};
    for (const [slug, r] of Object.entries(perInvestor)) {
      if (!r.ok) continue; // per-investor failure -> omit (UI treats missing as no-data)
      const raw = {
        ytd_pct: round4(r.ytd_pct),
        coverage_pct: round4(r.coverage_pct),
        priced_holdings: r.priced_holdings,
        eligible_holdings: r.eligible_holdings,
      };
      const withGate = applyCoverageGate(raw);
      gated[slug] = withGate.ytd_pct === null && raw.ytd_pct !== null;
      newInvestors[slug] = withGate;
    }

    const newUsable = usableCount(newInvestors);
    console.log(`[info] usable ytd_pct count: new=${newUsable} previous=${prevUsable === null ? "n/a" : prevUsable}`);

    const regression = checkRegression(newUsable, prevUsable);
    if (!regression.ok) {
      console.error(`::error::investor-returns ${regression.message}`);
      process.exitCode = 1;
      return;
    }
    if (regression.warn && regression.message) {
      console.warn(`::warning::investor-returns: ${regression.message}`);
    }

    const out = {
      as_of: nowIsoUtcMicros(),
      ytd_start: ytdStart,
      investors: newInvestors,
    };

    await fs.mkdir(path.dirname(OUTPUT_PATH), { recursive: true });
    await fs.writeFile(OUTPUT_PATH, JSON.stringify(out, null, 1) + "\n", "utf8");
    console.log(`[info] wrote ${OUTPUT_PATH}`);

    console.log("\nslug,ytd_pct,coverage_pct,gated");
    for (const [slug, r] of Object.entries(newInvestors)) {
      const ytd = r.ytd_pct === null ? "null" : (r.ytd_pct * 100).toFixed(2) + "%";
      const cov = (r.coverage_pct * 100).toFixed(1) + "%";
      console.log(`${slug},${ytd},${cov},${gated[slug] ? "yes" : "no"}`);
    }
  } finally {
    await browser.close();
  }
}

const isDirectRun = (() => {
  const entry = process.argv[1];
  if (!entry) return false;
  try {
    return fileURLToPath(import.meta.url) === path.resolve(entry);
  } catch {
    return false;
  }
})();

if (isDirectRun) {
  main().catch((e) => {
    console.error("[fail] uncaught error:", e);
    process.exitCode = 1;
  });
}
